consensus/types/src/data_column_sidecar.rs

```diff
@@ -62,94 +64,75 @@ pub struct DataColumnSidecar<T: EthSpec> {
+    }
 
-        (0..T::number_of_columns())
-            .map(|col_index| {
-                Ok(DataColumnSidecar {
-                    index: col_index as u64,
-                    column: Self::generate_column_data(&mut rng, num_of_blobs, col_index)?,
-                    kzg_commitments: blob_sidecars
-                        .iter()
-                        .map(|b| b.kzg_commitment)
-                        .collect::<Vec<_>>()
-                        .into(),
-                    kzg_proofs: blob_sidecars
-                        .iter()
-                        .map(|b| b.kzg_proof)
-                        .collect::<Vec<_>>()
-                        .into(),
-                    signed_block_header: first_blob_sidecar.signed_block_header.clone(),
-                    kzg_commitments_inclusion_proof: kzg_commitments_inclusion_proof.clone(),
-                })
-            })
-            .collect::<Result<Vec<_>, _>>()
+    pub fn block_root(&self) -> Hash256 {
+        self.signed_block_header.message.tree_hash_root()
     }
 
-    fn generate_column_data(
-        rng: &mut StdRng,
-        num_of_blobs: usize,
-        index: usize,
-    ) -> Result<DataColumn<T>, DataColumnSidecarError> {
-        let mut dummy_cell_data = Cell::<T>::default();
-        // Prefix with column index
-        let prefix = index.to_le_bytes();
-        dummy_cell_data
-            .get_mut(..prefix.len())
-            .ok_or(DataColumnSidecarError::DataColumnIndexOutOfBounds)?
-            .copy_from_slice(&prefix);
-        // Fill the rest of the vec with random values
-        rng.fill(
-            dummy_cell_data
-                .get_mut(prefix.len()..)
-                .ok_or(DataColumnSidecarError::DataColumnIndexOutOfBounds)?,
-        );
+    pub fn build_sidecars(
+        blobs: &BlobSidecarList<T>,
+        block: &SignedBeaconBlock<T>,
+        kzg: &Kzg,
+    ) -> Result<DataColumnSidecarList<T>, DataColumnSidecarError> {
+        let kzg_commitments = block
+            .message()
+            .body()
+            .blob_kzg_commitments()
+            .map_err(|_err| DataColumnSidecarError::PreDeneb)?;
+        let kzg_commitments_inclusion_proof =
+            block.message().body().kzg_commitments_merkle_proof()?;
+        let signed_block_header = block.signed_block_header();
 
-        let column = DataColumn::<T>::new(vec![dummy_cell_data; num_of_blobs])?;
-        Ok(column)
-    }
+        let mut columns =
+            vec![Vec::with_capacity(T::max_blobs_per_block()); T::number_of_columns()];
+        let mut column_kzg_proofs =
+            vec![Vec::with_capacity(T::max_blob_commitments_per_block()); T::number_of_columns()];
+
+        // NOTE: assumes blob sidecars are ordered by index
+        for blob in blobs {
+            let blob = KzgBlob::from_bytes(&blob.blob).map_err(KzgError::from)?;
+            let (blob_cells, blob_cell_proofs) = kzg.compute_cells_and_proofs(&blob)?;
```

Comment:
Would be useful adding some timer metrics for the cell and proofs computation.

Reply:
is there a convention for adding metrics? it appears that there are no metrics present in `consensus::types`.

Reply:
yeah I'm not aware of any conventions, but we try to be consistent with exiting code. 
We have timer to track inclusion proof inclusion time when building blob sidecars here:
https://github.com/sigp/lighthouse/blob/b047d3b690b4e5f3b08e9af05e12a4b03cb2f544/beacon_node/beacon_chain/src/block_verification.rs#L731-L735

Perhaps similarly we can do the same for data column sidecar and track how long it takes to build column sidecars?
https://github.com/sigp/lighthouse/blob/b047d3b690b4e5f3b08e9af05e12a4b03cb2f544/beacon_node/beacon_chain/src/block_verification.rs#L760

Reply:
Addressed in https://github.com/sigp/lighthouse/pull/5210/commits/96a9897d0ed9932bc6ce6a157a55c6cce95fe859